lib/std/fmt.zig

```diff
@@ -641,30 +649,57 @@ pub fn formatAsciiChar(
     return writer.writeAll(@as(*const [1]u8, &c));
 }
 
+pub fn formatUnicodeCodepoint(
+    c: u21,
+    options: FormatOptions,
+    writer: anytype,
+) !void {
+    if (unicode.utf8ValidCodepoint(c)) {
+        var buf: [4]u8 = undefined;
+        // The codepoint is surely valid, hence the use of unreachable
+        const len = std.unicode.utf8Encode(c, &buf) catch |err| switch (err) {
+            error.Utf8CannotEncodeSurrogateHalf, error.CodepointTooLarge => unreachable,
+        };
+        return formatBuf(buf[0..len], options, writer);
+    }
+
+    // In case of error output the replacement char U+FFFD
+    return formatBuf(&[_]u8{ 0xef, 0xbf, 0xbd }, options, writer);
```

Comment:
Why not just output the replacement char if `utf8Encode` returns an error?

Reply:
hm? `0xef, 0xbf, 0xbd` is the UTF-8 encoded replacement char

Reply:
I meant that `utf8Encode` returns an error for invalid input so there should be no need to validate before it?

Reply:
Oh right, GH won't let me see the lines around this.
Yes that's better, this check must be a leftover for some error-free `utf8Encode` alternative I was toying with.